### PR TITLE
fused dataset

### PR DESCRIPTION
**Pull request by tchaton**

### CODE REVIEW

File: conf/models/segmentation/pointnet2.yaml [hunk 1]
```diff
@@ -86,7 +86,8 @@ models:
     pointnet2_onehot:
         class: pointnet2.PointNet2_D
         conv_type: "DENSE"
-        use_category: True
+        define_constants:
```
Comment (nicolas-chaulet) [inline conf/models/segmentation/pointnet2.yaml]: This isn't enough, you also need to define a `use_category` key in the pointnet2_onehot object. Otherwise the model will not see it. (Same goes for the other models)

File: src/core/data_transform/transforms.py [hunk 2]
```diff
@@ -14,8 +16,91 @@
 
 from src.datasets.multiscale_data import MultiScaleData
 from src.utils.transform_utils import SamplingStrategy
+from src.utils.config import is_list
+from torch_geometric.data import Data, Batch
+from tqdm import tqdm as tq
 from src.utils import is_iterable
 
+class PointCloudFusion(object):
```
Comment (nicolas-chaulet) [inline src/core/data_transform/transforms.py]: A docstring would be nice here

File: src/core/data_transform/transforms.py [hunk 3]
```diff
@@ -14,8 +16,91 @@
...
+        return data
+
+    def __repr__(self):
+        return "{}()".format(self.__class__.__name__)
+
+class GridSphereSampling(object):
+    r"""Fit the point cloud to a grid and for each point in this grid, 
+    create a sphere with a radius r
+    Args:
+        radius (float or [float] or Tensor): Radius of the sphere to be sampled.
```
Comment (nicolas-chaulet) [inline src/core/data_transform/transforms.py]: Add the grid_size parameter and the transform returns (A list of data objects, one per sphere)

File: src/core/data_transform/transforms.py (hunk 2 above)
Comment (nicolas-chaulet) [inline src/core/data_transform/transforms.py]: If PointCloudFusion and GridSpehereSampling are pre collate transform then it might be a good idea to extract them in another file called `precollate_transforms.py`. This will reduce the risk of using those in the wrong way (as pre-transforms or transforms)

File: src/core/data_transform/transforms.py [hunk 4]
```diff
@@ -14,8 +16,91 @@
...
+        # The kdtree has bee attached to data for optimization reason.
+        # However, it won't be used for down the transform pipeline and should be removed before any collate func call.
+        if hasattr(data, self.KDTREE_KEY) and self._delattr_kd_tree:
+            delattr(data, self.KDTREE_KEY)
+
+        # apply grid sampling
+        grid_data = self._grid_sampling(data.clone())
+
+        datas = []
+        for grid_center, grid_label in zip(np.asarray(grid_data.pos), np.asarray(grid_data.y)):
```
Comment (nicolas-chaulet) [inline src/core/data_transform/transforms.py]: This logic will work only if the data object is not a batch. Could you add an assert to make sure that batch is not present in the data?

File: src/core/data_transform/transforms.py [hunk 5]
```diff
@@ -38,14 +130,15 @@ class RandomSphere(object):
         radius (float or [float] or Tensor): Radius of the sphere to be sampled.
```
Comment (nicolas-chaulet) [inline src/core/data_transform/transforms.py]: Can you please add a description of what this transform returns? Also explain the centre option

File: switch_cuda.sh [hunk 6]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
```
Comment (nicolas-chaulet) [inline switch_cuda.sh]: Not sure we want this file in our repo. Why did you add it?

File: src/datasets/samplers.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import torch
+import numpy as np
+from torch.utils.data import Sampler
+
+class BalancedRandomSampler(Sampler):
+    def __init__(self, labels, replacement=True):
```
Comment (nicolas-chaulet) [inline src/datasets/samplers.py]: A docstring would be nice

File: src/datasets/segmentation/s3dis.py [hunk 8]
```diff
@@ -131,6 +137,52 @@ def add_weights(dataset, train, class_weight_method):
...
 
+################################### 1m cylinder s3dis ###################################
+
+class S3DIS1x1Dataset(BaseDataset):
+    def __init__(self, dataset_opt, training_opt):
+        super().__init__(dataset_opt, training_opt)
+
+        pre_transform = self._pre_transform
+
+        transform = T.Compose(
```
Comment (nicolas-chaulet) [inline src/datasets/segmentation/s3dis.py]: Could you use the config file to set this transform?

File: src/datasets/segmentation/s3dis.py [hunk 9]
```diff
@@ -326,33 +378,193 @@ def get_tracker(model, task: str, dataset, wandb_opt: bool, tensorboard_opt: boo
...
+    def extract_center_labels(self, data_list):
+        extract_center_labels = []
+        for data in data_list:
+            if hasattr(data, "center_label"):
+                extract_center_labels.append(getattr(data, "center_label"))
+                delattr(data, "center_label")
+        return extract_center_labels
+
+    def save_center_labels(self, split_name, center_labels):
+        name = "{}_center_labels.p".format(split_name)
```
Comment (nicolas-chaulet) [inline src/datasets/segmentation/s3dis.py]: Please use .pt as a pytorch extension to be consistent

File: src/datasets/segmentation/s3dis.py [hunk 10]
```diff
@@ -326,33 +378,193 @@ def get_tracker(model, task: str, dataset, wandb_opt: bool, tensorboard_opt: boo
...
+               
+            torch.save(data_list, self.pre_processed_path)
+        else:
+            data_list = torch.load(self.pre_processed_path)
+
+        train_data_list = [data_list[i] for i in range(6) if i != self.test_area - 1]
+        test_data_list = data_list[self.test_area - 1]
+
+        if self.pre_collate_transform:
+            print('pre_collate_transform ...')
```
Comment (nicolas-chaulet) [inline src/datasets/segmentation/s3dis.py]: Use log instead of print please

File: src/core/data_transform/transforms.py (hunk 2 above)
Comment (tchaton) [reply src/core/data_transform/transforms.py]: I would prefer not too. They don't need to be used at this stage, and could be used as pre_transform.

Comment (tchaton) [reply src/core/data_transform/transforms.py]: I would prefer them not to be separated.

File: conf/models/segmentation/rsconv.yaml [hunk 11]
```diff
@@ -57,7 +57,8 @@ models:
     RSConv_MSN:
         class: rsconv.RSConvLogicModel
         conv_type: "DENSE"
-        use_category: True
+        define_constants:
```
Comment (nicolas-chaulet) [inline conf/models/segmentation/rsconv.yaml]: My point was that you need to define a `use_category` member inside each model. Otherwise it gets missed.

File: src/datasets/segmentation/s3dis.py (hunk 8 above)
Comment (nicolas-chaulet) [reply src/datasets/segmentation/s3dis.py]: transform is not used now, could you please remove it?
